pydough/errors/pydough_error_builder.py

```diff
@@ -0,0 +1,344 @@
+    into the session to customize the error messages.
+    """
+
+    def term_not_found(
+        self, collection: "PyDoughCollectionQDAG", term_name: str
+    ) -> PyDoughException:
+        """
+        Creates an exception for when a term is not found in the specified collection.
+
+        Args:
+            `collection`: The collection in which the term was not found.
+            `term_name` The name of the term that was not found.
+        Returns:
+            An exception indicating that the term was not found.
+        """
+        return PyDoughQDAGException(
+            collection.name_mismatch_error(
+                term_name,
+                atol=2,
+                rtol=0.1,
+                min_names=3,
+                insert_cost=0.5,
+                delete_cost=1.0,
+                substitution_cost=1.0,
+                capital_cost=0.1,
+            )
+        )
+
+    def down_streaming_conflict(
+        self, collection: "PyDoughCollectionQDAG", term_name: str
+    ) -> PyDoughException:
+        """
+        Creates an exception for when a term accessed within a collection but
+        it is unclear whether it is a term of the collection or a term
+        downstreamed from an ancestor.
+
+        Args:
+            `collection`: The collection in which the term is being accessed.
+            `term_name`: The name of the term that caused the ambiguity.
+
+        Returns:
+            An exception indicating the name access ambiguity.
+        """
+        return PyDoughQDAGException(
+            f"Unclear whether {term_name!r} refers to a term of the current context or ancestor of collection {collection!r}"
+        )
+
+    def cardinality_error(
+        self, collection: "PyDoughCollectionQDAG", expr: "PyDoughExpressionQDAG"
+    ) -> PyDoughException:
+        """
+        Creates an exception for when a term is used within a context that
+        should be singular with regards to the context, but it is plural.
+
+        Args:
+            `collection`: The collection in which the term is being accessed.
+            `expr`: The PyDoughQDAG expression
+
+        Returns:
+            An exception indicating the cardinality error.
+        """
+        raise PyDoughQDAGException(
+            f"Expected all terms in {collection.standalone_string} to be singular, but encountered a plural expression: {expr}"
+        )
+
+    def expected_collection(self, expr: object) -> PyDoughException:
+        """
+        Creates an exception for when a QDAG collection is expected but
+        something else is found.
+        """
+        from pydough.qdag import PyDoughExpressionQDAG
+
+        if isinstance(expr, PyDoughExpressionQDAG):
+            return PyDoughQDAGException(
+                f"Expected a collection, but received an expression: {expr}"
+            )
+        else:
+            return PyDoughQDAGException(
+                f"Expected a collection, but received {expr.__class__.__name__}:  {expr}"
```

Comment:
Can we explicitly handle the `None` case. I remember getting message like `Expected a collection, but received NoneType: None` which is confusing
Suggestion
```
if expr is None:
    return PyDoughQDAGException("Expected a collection, but received None")
```